src/common/perf_counters_cache_key.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+#include <algorithm>
+#include <initializer_list>
+#include <iterator>
+#include <numeric>
+#include <string_view>
+#include <utility>
+
+namespace ceph::perf_counters {
+
+/// key/value pair representing a perf counter label
+using label_pair = std::pair<std::string_view, std::string_view>;
+
+/// construct a cache key for a perf counter and set of labels. returns a string
+/// of the form "counter_name\0key1\0value1\0key2\0value2\0", where label pairs
+/// are insertion-sorted by key then value
+constexpr std::string make_cache_key(std::string_view counter_name,
+                                     std::initializer_list<label_pair> labels);
+
+// TODO: cache_key_insert() to add more labels to an existing string
+
+
+namespace detail {
+
+// use a null character to delimit strings
+constexpr char delimiter = '\0';
+
+// write a string and null delimiter to the output
+constexpr auto insert(std::string_view str,
+                      std::output_iterator<char> auto out)
+{
+  out = std::copy(str.begin(), str.end(), out);
+  *out = delimiter;
+  return std::next(out);
+}
+
+constexpr auto parse_label(std::contiguous_iterator auto pos,
```

Comment:
The first concepts in Ceph? Cool!

Reply:
this approach iteratively reformats the key as a buffer?  is there not an equivalent that yields a sequence of strings in the defined order?

Reply:
> this approach iteratively reformats the key as a buffer?

right, this sorting algorithm only uses that flat buffer for storage. i started with this simple insertion sort to show that it could be done with the single std::string allocation

> is there not an equivalent that yields a sequence of strings in the defined order?

we could allocate a temporary array of `label_pair`s and sort that first - then `parse_label()` wouldn't need to search the buffer for delimiters, and `insert_label()` wouldn't need to shift memory around for out-of-order inserts

if you guys think this string-based strategy is worth pursuing, i'm happy to put some more effort into benchmarking and optimization here. it is important that we optimize this path, because ceph daemons expect that their use of perf counters don't have a significant effect on the performance they're trying to measure

Reply:
I definitely think this is worth pursuing. I was thinking of something along similar lines last night.

Reply:
Those ideas sound interesting.  It did occur to me that we should consider using a differently-constructed prototype for the incoming "find-counter-with-labels" comparisons.  i.e., the version of the key stored with a given counter in the cache would be allocated, but the prototype key used for comparison could just use string_views from the environment.  In  rgwnfs filehandle comparison, I used a function object as the prototype, but I guess we now have lots of ways to construct one.